### DISCUSSION TITLE
Collapse/Expand multiline logs

### DISCUSSION THREAD

**Discussion by ngc104** [grafana/grafana#45802] [Dashboards, Panels & Visualizations]
## TL;DR

Add a button to expand/collapse multiline logs (without the help of "Escape newlines" button).

## Context

I have multiline logs (stacktraces...) that I can render as one log thanks to promtail multiline stage.

In Grafana, I can see my logs spreading over multiple lines.

However, there are some times when I would like to put them into one line (one line per timestamp) in the same way as the "wrap lines" button does for single lines logs.

## What I have

I wrote a little program that send a short stacktrace in the log (`python -c "print(1/0)"`). The stacktrace will spread over 3 lines. But you can easily imagine a long stacktrace as Java can do.

Here is a simple expression to retrieve my logs in my dedicated namespace :

```
{namespace="multiline-grafana-test"}
```

![1](https://user-images.githubusercontent.com/67380650/155372375-984d5aa0-c6ec-465e-ab72-e445cd16daf8.png)

## How I can do what I want

In order to put those 3 lines into one, I can improve my logQL line :

```
{namespace="multiline-grafana-test"} | regexp "^(?P<msg>(?s:.*))$" | line_format "{{ regexReplaceAll \"\\n\" .msg \"\\\\n\"}}"
```

I guess that with a more recent version of Loki, I could even do this :

```
{namespace="multiline-grafana-test"} | line_format "{{ regexReplaceAll \"\\n\" .message \"\\\\n\"}}"
```

![2](https://user-images.githubusercontent.com/67380650/155372420-a7553c59-0639-48b0-bdf0-bfd55c6544d6.png)

As you can see in the screenshots, Grafana will detect the `\n` characters and show me a button `Escape Newlines`. When I clic on this button, I get exactly what I want.

![3](https://user-images.githubusercontent.com/67380650/155372454-3a1493a1-acfc-4642-888c-959b96a3e311.png)

## Feature request

The process to get multilines expanded/collapsed is too complex. Could you do a more user-friendly process with just a button ?

**Suggestion 1** : Grafana could show a button just like "Wrap lines" or "Prettify Json". That could be "Expand multiple lines" or something similar. Grafana could then detect the "end of line" character (the one I replaced with the `\n` string) and :
- when the user enable "Expand multiple lines", display the expanded log (as it does it already)
- when the user disables "Expand multiple lines", replace all the "end of line" character with a special character  

Pros : intuitive and easy to find
Cons : global

**Suggestion 2** : near the left arrow (and the time that can be shown or hidden with the "Time" button), you could add a button there to expand/collapse the multilines. Or that could be on the right of the log.

Pros : let the user expand only the logs he wants
Cons : many buttons at the left or the right of the logs

**Environment**:
- Grafana version: 8.1.6
- Loki: 2.2.1
- OS Grafana is installed on: Kubernetes
- User OS & Browser: Any

**Comment by svennergr** [comment]
Please use https://github.com/grafana/grafana/issues/74261 for further discussion. Thank you!
